clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -155,7 +155,7 @@ private enum State {
         private boolean isTransitionValid(State source, State target) {
             switch (target) {
                 case UNINITIALIZED:
-                    return source == READY;
+                    return source == READY || source == ABORTABLE_ERROR;
```

Comment:
Do we know if there are other implications for this state machine change? It seems sort of like we want to just make the authorization error retriable -- but instead we are manking it abortable + making this specific abortable error reset the producer state.

I'm not an expert in this area, but are there any other implications with re-initializing the producer?

Reply:
It's one way (that I could think of) to propagate the authorization error from the sender loop; otherwise, the sender will continue to retry and causes timeout in some cases.

Reply:
Hmmm. I see. I'm just wondering if we are hijacking the current state machine in an unexpected way and if there are implications there. I suppose we are only following this path on these specific error types, but I wonder if we are missing anything existing by changing the valid transitions and/or opening up the potential for something else in the future.

Reply:
hmm good point, I guess upon re-initializing (transition from UNINITIALIZED to INITIALIZING state), should we check the previous error to ensure a valid transition? Maybe in `initializeTransactions` we examine the previous error and make the next transition?

Reply:
Can you point to me where you saw this? I think abortable error was something we had before that didn't need this extra step.

Reply:
We need to manually transition the state back to initializing no? I think otherwise it get stucks in an error state here

https://github.com/apache/kafka/blob/a2c9f421af40e0c8ace722be94aedf8dec4f2b31/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L998

Reply:
Ack -- I wonder if there are any other location where we would not want this transition. I can't think of off the top of my head, but I do think we need to be careful to clean the state properly in these cases.